src/prefect_server/api/flow_concurrency_limits.py

```diff
@@ -0,0 +1,162 @@
+import asyncio
+from typing import Dict, List, Optional
+
+import prefect
+from prefect import models
+from prefect.engine.state import Running, Submitted, State
+from prefect.utilities.plugins import register_api
+
+OCCUPYING_STATES = [
+    state.__name__
+    for state in prefect.engine.state.__dict__.values()
+    if isinstance(state, type) and issubclass(state, (Running, Submitted))
+]
+
+
+@register_api("flow_concurrency_limits.update_flow_concurrency_limit")
+async def update_flow_concurrency_limit(tenant_id: str, name: str, limit: int) -> str:
+    """
+    Updates or creates a flow concurrency limit.
+
+    Args:
+        - tenant_id (str): The tenant owner of the limit
+        - name (str): The name of the limit
+        - limit (int): The maximum number of concurrency flows allowed
+
+    Returns:
+        - str: The ID of the created or updated flow concurrency limit
+
+    Raises:
+        - ValueError: If the `limit` is invalid
+        - ValueError: If the limit exists and updating it fails
+    """
+
+    if limit <= 0:
+        raise ValueError(
+            (
+                "Flow concurrency limits require positive integers representing"
+                f" the maximum number of concurrent allowed flows. Got {limit}"
+            )
+        )
+
+    existing = await models.FlowConcurrencyLimit.where(
+        {"name": {"_eq": name}, "tenant_id": {"_eq": tenant_id}}
+    ).first({"id", "name", "limit"})
+
+    if not existing:
+        return await models.FlowConcurrencyLimit(
+            name=name, limit=limit, tenant_id=tenant_id
+        ).insert()
+
+    if not existing.limit == limit:
+        result = await models.FlowConcurrencyLimit.where(id=existing.id).update(
+            set=dict(limit=limit)
+        )
+        if not bool(result.affected_rows):
+            raise ValueError("Error updating existing flow concurrency limit")
+
+    return existing.id
+
+
+@register_api("flow_concurrency_limits.delete_flow_concurrency_limit")
+async def delete_flow_concurrency_limit(limit_id: str) -> bool:
+    """
+    Deletes a flow concurrency limit.
+
+    Args:
+        - limit_id (str): The flow concurrency limit to delete.
+
+    Returns:
+        - bool: If the delete was successful
+
+    Raises:
+        - ValueError: If an ID isn't provided
+    """
+
+    if not limit_id:
+        raise ValueError("Invalid flow concurrency limit ID.")
+
+    result = await models.FlowConcurrencyLimit.where(id=limit_id).delete()
+    return bool(result.affected_rows)
+
+
+@register_api("flow_concurrency_limits.try_take_flow_concurrency_slots")
+async def try_take_flow_concurrency_slots(
+    tenant_id: str, limit_names: List[str], current_state: Optional[State] = None
+) -> bool:
```

Comment:
The previous signature took a `flow_run_id: str = None` parameter instead of `current_state: Optional[State] = None`. This was changed as all places this check is done already have queried the flow run and have existing state data available. For compatibility concerns, this can be changed back, but will likely slow down performance.